mp2-v1/src/receipt_extraction/leaf.rs

```diff
@@ -0,0 +1,471 @@
+//! Module handling the leaf node inside a Receipt Trie
+
+use crate::MAX_RECEIPT_LEAF_NODE_LEN;
+
+use super::public_inputs::{PublicInputArgs, PublicInputs};
+
+use mp2_common::{
+    array::{Array, Vector, VectorWire},
+    eth::{EventLogInfo, LogDataInfo, ReceiptProofInfo},
+    group_hashing::CircuitBuilderGroupHashing,
+    keccak::{InputData, KeccakCircuit, KeccakWires},
+    mpt_sequential::{MPTReceiptLeafNode, ReceiptKeyWire, MAX_TX_KEY_NIBBLE_LEN, PAD_LEN},
+    public_inputs::PublicInputCommon,
+    types::{CBuilder, GFp},
+    utils::{Endianness, PackerTarget},
+    D, F,
+};
+use plonky2::{
+    field::types::Field,
+    iop::{
+        target::Target,
+        witness::{PartialWitness, WitnessWrite},
+    },
+    plonk::circuit_builder::CircuitBuilder,
+};
+
+use plonky2_ecgfp5::gadgets::curve::{CircuitBuilderEcGFp5, CurveTarget};
+
+use recursion_framework::circuit_builder::CircuitLogicWires;
+use rlp::Encodable;
+use serde::{Deserialize, Serialize};
+use std::array::from_fn;
+/// Maximum number of logs per transaction we can process
+const MAX_LOGS_PER_TX: usize = 2;
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ReceiptLeafWires<const NODE_LEN: usize>
+where
+    [(); PAD_LEN(NODE_LEN)]:,
+{
+    /// The event we are monitoring for
+    pub event: EventWires,
+    /// The node bytes
+    pub node: VectorWire<Target, { PAD_LEN(NODE_LEN) }>,
+    /// the hash of the node bytes
+    pub root: KeccakWires<{ PAD_LEN(NODE_LEN) }>,
+    /// The index of this receipt in the block
+    pub index: Target,
+    /// The offset of the status of the transaction in the RLP encoded receipt node.
+    pub status_offset: Target,
+    /// The offsets of the relevant logs inside the node
+    pub relevant_logs_offset: VectorWire<Target, MAX_LOGS_PER_TX>,
+    /// The key in the MPT Trie
+    pub mpt_key: ReceiptKeyWire,
+}
+
+/// Contains all the information for an [`Event`] in rlp form
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct EventWires {
+    /// Size in bytes of the whole event
+    size: Target,
+    /// Packed contract address to check
+    address: Array<Target, 20>,
```

Comment:
We have many types defined here https://github.com/Lagrange-Labs/mapreduce-plonky2/blob/main/mp2-common/src/types.rs - such as a type for an Address.